src/wayland/seat/touch.rs

```diff
@@ -109,83 +129,104 @@ impl TouchInternal {
+        self.with_focused_handles(dh, slot, |dh, handle| {
+            if handle.version() >= 6 {
+                handle.shape(dh, slot.into(), major, minor);
             }
         });
     }
 
-    fn orientation(&self, slot: TouchSlot, orientation: f64) {
-        self.with_focused_handles(slot, |handle| {
-            if handle.as_ref().version() >= 6 {
-                handle.orientation(slot.into(), orientation);
+    fn orientation(&self, dh: &mut DisplayHandle<'_>, slot: TouchSlot, orientation: f64) {
+        self.with_focused_handles(dh, slot, |dh, handle| {
+            if handle.version() >= 6 {
+                handle.orientation(dh, slot.into(), orientation);
             }
         });
     }
 
     // TODO: In theory doesn't need to be sent for WlTouch that isn't in the focus hashmap?
-    fn cancel(&self) {
+    fn cancel(&self, dh: &mut DisplayHandle<'_>) {
         for handle in &self.known_handles {
-            handle.cancel();
+            handle.cancel(dh);
         }
     }
 
     // TODO: Document this also sends frame every time.
     #[inline]
-    fn with_focused_handles<F>(&self, slot: TouchSlot, mut f: F)
+    fn with_focused_handles<F>(&self, dh: &mut DisplayHandle<'_>, slot: TouchSlot, mut f: F)
     where
-        F: FnMut(&WlTouch),
+        F: FnMut(&mut DisplayHandle<'_>, &WlTouch),
     {
         if let Some(focus) = self.focus.get(&slot) {
             for handle in &focus.handles {
-                f(handle);
-                handle.frame();
+                f(dh, handle);
+                handle.frame(dh);
             }
         }
     }
 }
 
-pub(crate) fn implement_touch(touch: Main<WlTouch>, handle: Option<&TouchHandle>) -> WlTouch {
-    // The sole `Release` request is already handled by our destructor.
-    touch.quick_assign(|_touch, _request, _data| {});
+/// User data for keyboard
+#[derive(Debug)]
+pub struct TouchUserData {
+    pub(crate) handle: Option<TouchHandle>,
+}
 
-    // Remove from touch handles on destroy.
-    if let Some(handle) = handle {
-        let inner = handle.inner.clone();
-        touch.assign_destructor(Filter::new(move |touch: WlTouch, _, _| {
-            inner
-                .borrow_mut()
-                .known_handles
-                .retain(|t| !t.as_ref().equals(touch.as_ref()))
-        }));
+impl<T> DelegateDispatchBase<WlTouch> for SeatState<T> {
+    type UserData = TouchUserData;
+}
+
+impl<T, D> DelegateDispatch<WlTouch, D> for SeatState<T>
+where
+    D: 'static + Dispatch<WlTouch, UserData = TouchUserData>,
+    D: SeatHandler<T>,
+{
+    fn request(
+        _state: &mut D,
+        _client: &wayland_server::Client,
+        _resource: &WlTouch,
+        _request: wl_touch::Request,
+        _data: &Self::UserData,
+        _dhandle: &mut DisplayHandle<'_>,
+        _data_init: &mut wayland_server::DataInit<'_, D>,
+    ) {
+        unreachable!("There should be no wl_touch requests");
```

Comment:
Sorry I have made a mistake, there is a destructor. Let's just put a comment about that.